### PR TITLE
feat(MultiThreadedExecutor): Added ability to handle exceptions from threads

### PR DESCRIPTION
**Pull request by jmachowinski**
This commit adds external exception handling for the worker threads, allowing application code to implement custom exception handling.

### CODE REVIEW

**Comment by jmachowinski** [comment]
@wjwwood  @mjcarroll  @fujitatomoya Ping, I would like to see this merged for jazzy

File: rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: why this is dedicated to `MultiThreadedExecutor`? i think user application can find useful with other Executors as well.

File: rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp:60]: how about adding the test case for this to make sure the callback suppresses the exception?

**Review by fujitatomoya — COMMENTED** [review]
@jmachowinski thanks for the PR. i think this is useful! a couple of minor comments. (if we merge this, i think it would be nice to update the demos/example aligned with this.)

File: rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp
Comment (alsora) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: In a single thread executor, the user is the one that is creating the thread where we spin, so they can take care of handling the exception without any new API.
This is not the case for the threads spawned internally by an executor, such as in this case

Comment (fujitatomoya) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: yeah i understand that, but user can use the same manner for any executor? i just though that would be useful to write the application.

Comment (jmachowinski) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: I don't care either way...

File: rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp
Comment (jmachowinski) [reply rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp:60]: Added test case

File: rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp
Comment (jmachowinski) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: @fujitatomoya @alsora can you come to an agreement, so that we can move forward with this PR ?

Comment (alsora) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:82]: I'm ok with adding it to all executors.

Comment (jmachowinski) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp]: Added spin API with exception handler to Executor class

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI.

@alsora friendly ping.

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp
Comment (alsora) [inline rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:102]: can this test be moved to the "common executors tests" ?

File: rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp
Comment (alsora) [inline rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp:95]: to be precise, the events executor should have a dedicated implementation.
This line here https://github.com/ros2/rclcpp/blob/070d71e2c68bd8009c7f96c3c110f2cb2390dde3/rclcpp/src/rclcpp/experimental/executors/events_executor/events_executor.cpp#L114 will start a new thread

Comment (jmachowinski) [reply rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp:95]: fair point...

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp
Comment (jmachowinski) [reply rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:102]: I'll look into it...

Comment (jmachowinski) [reply rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp]: done

File: rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp
Comment (jmachowinski) [reply rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp]: done

**Comment by fujitatomoya** [comment]
@alsora can you check and resolve your comments above?

**Comment by fujitatomoya** [comment]
@jmachowinski can you rebase to rolling? i would like to make sure CI is green with current HEAD especially at this release moment.

**Comment by alsora** [comment]
Looks good to me

**Review by alsora — APPROVED** [review]

**Comment by jmachowinski** [comment]
> @jmachowinski can you rebase to rolling? i would like to make sure CI is green with current HEAD especially at this release moment.
> 
rebased, note this currently fails the pipline, as it uncovered the bug reported in https://github.com/ros2/rclcpp/issues/2474

**Comment by fujitatomoya** [comment]
@jmachowinski this is ready for CI with https://github.com/ros2/demos/pull/670? just chekcing.

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20841/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15144/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21527/)

**Comment by jmachowinski** [comment]
rebased to rolling

I don't know why the CI failed, it builds for me using this file:
https://github.com/jmachowinski/build_desc/raw/exeception_handling/ros2.repos

**Comment by wjwwood** [comment]
I need to make some changes before merging this, but I cannot push to this branch, so I'll be opening a new pr. Please don't merge this one.

File: rclcpp/include/rclcpp/executor.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp:95]: I would have liked to have had a non-virtual `spin(... exception_handler = nullptr)`, that called a virtual and protected `spin_impl(... exception_handler) = 0`, so that we didn't need two signatures. It would be easier to document and wouldn't require every executor to implement two functions, one of which is pretty much always be boilerplate calling the other.

However, I don't know if I have time to do that before the freeze, so I won't block this on it, but I wanted to make the comment somewhere.

File: rclcpp/include/rclcpp/executor.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp:88]: We don't use `\` everywhere (some other `@` have slipped through reviews), but we should tend towards the desired state when possible, which is `\` everywhere. I'll fix these up in my version of this pr.

File: rclcpp/include/rclcpp/executor.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp:86]: This needs a "brief", with a `/// ...` line before the block comment.

File: rclcpp/include/rclcpp/experimental/timers_manager.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/experimental/timers_manager.hpp:129]: I'll leave this `@`, because this file uses that style, but the whole file should be updated (it mostly got a review pass for being in the `experimental` namespace.

File: rclcpp/src/rclcpp/executor.cpp
Comment (wjwwood) [inline rclcpp/src/rclcpp/executor.cpp:400]: nitpick: use full variable names

```suggestion
  const Function & function,
```

File: rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp
Comment (wjwwood) [inline rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp:63]: We should not copy-paste like this, we should have one call the other...

File: rclcpp_lifecycle/test/benchmark/benchmark_lifecycle_client.cpp
Comment (wjwwood) [inline rclcpp_lifecycle/test/benchmark/benchmark_lifecycle_client.cpp:206]: Why is this change needed?

**Review by wjwwood — CHANGES_REQUESTED** [review]
I am going to fix up this pr and run CI on it as an option to merge, but I really think that it would be preferable to have an exception handler callback as an optional part of the ExecutorOptions instead, passing it during construction and/or setting it after construction with a new method, rather than modifying all of these spin functions and causing down stream changes.

Please don't merge this version of the pr, see: https://github.com/ros2/rclcpp/pull/2505

**Comment by fujitatomoya** [comment]
> but I really think that it would be preferable to have an exception handler callback as an optional part of the ExecutorOptions instead

+1 for this.

**Comment by wjwwood** [comment]
Another concern I've got after trying to shape this up for merge is that there's no way to indicate if an exception should cause the executor to stop or not. You can re-throw, but doing that in a thread will trigger termination, which by default will head towards `std::abort`, but can be overridden see: https://en.cppreference.com/w/cpp/error/set_terminate

Instead, I'd like to have seen either a return value from the exception handler that can tell the executor to stop or throw from the main thread or something, and/or pass the executor into the exception handler so that it can cancel it if desired.

File: rclcpp_lifecycle/test/benchmark/benchmark_lifecycle_client.cpp
Comment (wjwwood) [reply rclcpp_lifecycle/test/benchmark/benchmark_lifecycle_client.cpp:206]: I see it doesn't compile... this is the kind of change that suggests we should rethink the approach. The old style of calling it should still work I think. I'll leave it, but it's another point against this approach in my mind.

**Comment by wjwwood** [comment]
Closing in favor of the updated version https://github.com/ros2/rclcpp/pull/2505

**Pull request closed without merging**
